flang/lib/Lower/CharacterExpr.cpp

```diff
@@ -632,3 +645,30 @@ bool Fortran::lower::CharacterExprHelper::isArray(mlir::Type type) {
   }
   return false;
 }
+
+mlir::Value
+Fortran::lower::CharacterExprHelper::getLen(const fir::ExtendedValue &exv) {
+  // If the ExtendedValue has a length field, use it.
+  if (const auto *charBox = exv.getCharBox())
+    return charBox->getLen();
+  if (const auto *charArrayBox = exv.getBoxOf<fir::CharArrayBoxValue>())
+    return charArrayBox->getLen();
+  if (const auto *boxValue = exv.getBoxOf<fir::BoxValue>())
+    return boxValue->getLen();
+  if (auto *unboxed = exv.getUnboxed())
+    return toDataLengthPair(*unboxed).getLen();
+  // TODO: Is length of character procedure pointer needed ?
+  llvm::report_fatal_error("BoxProc length has not length");
+}
+
+fir::ExtendedValue
+Fortran::lower::CharacterExprHelper::cleanUpCharacterExtendedValue(
+    const fir::ExtendedValue &exv) {
+  if (const auto *charBox = exv.getCharBox())
```

Comment:
ExtendedValues support match syntax now. (Provides them with a more functional pattern matching syntax, to look like e.g. OCaml.)